tests/interpreter/cases/call/or.yaml

```diff
@@ -0,0 +1,114 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+cases:
+  - note: or-error
+    data: {}
+    modules:
+      - |
+        package test
+
+        inc(x) = x + 1
+        inc(x) = x + 4
+
+        a1 = inc(5)
+    query: data.test
+    error: functions must not produce multiple outputs for same inputs
+
+  - note: or-ok
+    data: {}
+    modules:
+      - |
+        package test
+
+        inc(x) = x + 1
+        inc(x) = x + 1
+
+        a1 = inc(5)
+    query: data.test
+    want_result:
+      a1: 6
+
+  - note: or-true
+    data: {}
+    modules:
+      - |
+        package test
+
+        inc(x) = 7 == x
+        inc(x) = 7 == x
+
+        a1 = inc(7)
+    query: data.test
+    want_result:
+      a1: true
+
+  - note: or-false
+    data: {}
+    modules:
+      - |
+        package test
+
+        inc(x) = 8 == x
+        inc(x) = 3 == x
+
+        a1 = inc(2)
+    query: data.test
+    want_result:
+      a1: false
+
+  - note: or-undefined
+    data: {}
+    modules:
+      - |
+        package test
+
+        inc(x) = x + 1
+        inc(x) = x + 1
+        inc(x) = y {
+           x > 10 # This will evaluate to false.
+           y = 100 # y will be undefined.
+        }
+
+        a1 = inc(5)
+    query: data.test
+    want_result:
+      a1: 6
+
+  - note: or-first-undefined
+    data: {}
+    modules:
+      - |
+        package test
+
+        inc(x) = y {
+           x > 10 # This will evaluate to false.
+           y = 100 # y will be undefined.
+        }
+        inc(x) = x + 1
+        inc(x) = x + 1
+
+        a1 = inc(5)
+    query: data.test
+    want_result:
+      a1: 6
+      
+
+  - note: or-all-undefined
+    data: {}
+    modules:
+      - |
+        package test
+
+        inc(x1) = y1 {
+           x1 > 10 # This will evaluate to false.
+           y1 = 100 # y will be undefined.
+        }
+        inc(x2) = y2 {
+           x2 > 10 # This will evaluate to false.
+           y2 = 100 # y will be undefined.
+        }
+
+        a1 = inc(5)
+    query: data.test
+    want_result: {}
```

Comment:
There is one more case to handle:
```
fcn(x) = y {
  y = x + 1
}

fcn(x) = y {
   y = concat(" ", ["hello", x])
}

a1 = fcn("world")
a2 = fcn(5)
```

This produces:
```
{
    "a1": "hello world",
    "a2": 6
}
```

The definitions that produce an error are ignored. If all definitions produce errors, then one of the errors is propagated.

Reply:
I added the test case without being able to fix it in this PR. 
I left a comment to not forget about it.

Unless I missed a simple solution, there is a lot of code to rewrite to support that ? especially how errors are handled.

Reply:
For minimal changes, we could slightly tweak your current logic:

1. Also push errors into a vector `errors` within the loop.
2. Then
   - If `results` is not empty then ignore `errors` and return the result after asserting that all the values are same.
   - If `results` is empty, then 
       - If `errors` is not empty, then return the first error.
       - If `errors` is empty, then return Undefined.

Reply:
I didn't know that errors are propagated like that in Rust ... thanks for the hint, it's super clear.

I still had a problem, the test case `or-one-error` returns
```
{
  "a1": [
    "hello world"
  ],
  "a2": [
    6
  ]
}
```

but in rego playground it is:
```
{
    "a1": "hello world",
    "a2": 6
}
```

I didn't investigate why this difference.

Reply:
Thanks for pointing it out. I will take a look at it.